src/dbnode/storage/namespace.go

```diff
@@ -1602,3 +1604,171 @@ func (n *dbNamespace) FlushState(shardID uint32, blockStart time.Time) (fileOpSt
 func (n *dbNamespace) nsContextWithRLock() namespace.Context {
 	return namespace.Context{ID: n.id, Schema: n.schemaDescr}
 }
+
+func (n *dbNamespace) AggregateTiles(
+	ctx context.Context,
+	sourceNs databaseNamespace,
+	opts AggregateTilesOptions,
+	pm persist.Manager,
+) (int64, error) {
+	callStart := n.nowFn()
+	processedBlockCount, err := n.aggregateTiles(ctx, sourceNs, opts, pm)
+	n.metrics.writeAggData.ReportSuccessOrError(err, n.nowFn().Sub(callStart))
+
+	return processedBlockCount, err
+}
+
+func (n *dbNamespace) aggregateTiles(
+	ctx context.Context,
+	sourceNs databaseNamespace,
+	opts AggregateTilesOptions,
+	pm persist.Manager,
+) (int64, error) {
+	targetBlockSize := n.Metadata().Options().RetentionOptions().BlockSize()
+	blockStart := opts.Start.Truncate(targetBlockSize)
+	if blockStart.Add(targetBlockSize).Before(opts.End) {
+		return 0, fmt.Errorf("tile aggregation must be done within a single target block (start=%s, end=%s, blockSize=%s)",
+			opts.Start, opts.End, targetBlockSize.String())
+	}
+
+	n.RLock()
+	if n.bootstrapState != Bootstrapped {
+		n.RUnlock()
+		return 0, errNamespaceNotBootstrapped
+	}
+	nsCtx := n.nsContextWithRLock()
+	n.RUnlock()
+
+	targetShards := n.OwnedShards()
+
+	// Note: Cold writes must be enabled for Large Tiles to work.
+	if !n.nopts.ColdWritesEnabled() {
+		return 0, errColdWritesDisabled
+	}
+
+	sourceBlockSize := sourceNs.Metadata().Options().RetentionOptions().BlockSize()
+	sourceBlockStart := opts.Start.Truncate(sourceBlockSize)
+
+	sourceNsOpts := sourceNs.StorageOptions()
+	reader, err := fs.NewReader(sourceNsOpts.BytesPool(), sourceNsOpts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		return 0, err
+	}
+
+	wOpts := series.WriteOptions{
+		TruncateType: n.opts.TruncateType(),
+		SchemaDesc:   nsCtx.Schema,
+	}
+
+	resources, err := newColdFlushReuseableResources(n.opts)
+	if err != nil {
+		return 0, err
+	}
+
+	// NB(bodu): Deferred targetShard cold flushes so that we can ensure that cold flush index data is
+	// persisted before persisting TSDB data to ensure crash consistency.
+	multiErr := xerrors.NewMultiError()
+	var processedBlockCount int64
+	for _, targetShard := range targetShards {
```

Comment:
May be good to get some further (non-trace) instrumentation around here, e.g. log time taken for completing this loop?

Reply:
Adding it on a later branch (because the code has really diverged here): https://github.com/m3db/m3/commit/aecae805908de6fa7c96f8860695575075f2f233